### PR TITLE
Dockerimage build + python dependencies via poetry

### PR DESCRIPTION
**Pull request by sandhawalia**
We use VPF extensively within our projects (classical vision + DL) and thought this might come handy for folks looking to explore VPF. Python dependencies are installed via `pyproject.toml` file using [python-poetry](https://python-poetry.org)

## Usage 

```
# Skip this if you have all the deps and docker/docker-compose installed
. scripts/install_deps.sh
. scripts/install_docker.sh
```

### Build image
```
export VIDEO_CODEC_SDK=<path_to_your_Video_Codec_SDK>
cp -a $(VIDEO_CODEC_SDK) Video_Codec_SDK
docker-compose -f docker/docker-compose.yml build vpf
```

### Run image
```
# Get test sample
wget http://www.scikit-video.org/stable/_static/bikes.mp4 $HOME/Downloads/
# launch image 
docker-compose -f docker/docker-compose.yml run vpf
# or run image with mounting a volume for testing (recommended)
docker run --runtime=nvidia -e NVIDIA_DRIVER_CAPABILITIES=video,compute,utility -v $HOME/Downloads:/Downloads -it nvidia/videoprocessingframework:vpf
python Tests.py 0 /Downloads/bikes.mp4 /Downloads/bikes-out.mp4
```

## Environment

Tested on Ubuntu 20:04 with Titan RTX GPU

```
lsb_release -a

Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
```

```
nvidia-smi -a

==============NVSMI LOG==============

Timestamp                                 : Thu Aug 26 18:23:40 2021
Driver Version                            : 460.32.03
CUDA Version                              : 11.2

Attached GPUs                             : 1
GPU 00000000:D8:00.0
    Product Name                          : TITAN RTX
```

NB:

1. We've also forked OpenCV and wrote glue code to convert [`SurfacePlane`](https://github.com/NVIDIA/VideoProcessingFramework/blob/master/PyNvCodec/src/PyNvCodec.cpp#L1670) to [`cv2.cuda.GpuMat`](https://docs.opencv.org/4.5.3/d0/d60/classcv_1_1cuda_1_1GpuMat.html) with zero copy. This facilitates using VPF in junction with OpenCV enabling accelerated CV vision applications. Would this be of interest to the folks using VPF?

### CODE REVIEW

**Comment by rarzumanyan** [comment]
@philipp-schmidt
Could you please give me a hand and share your opinion on this PR?

@sandhawalia 
I've found 2 possible issues we need to address:
- It looks like patch is compatible with Linux only (VPF supports Windows as well).
- Makefile is located at the project's top level directory which may cause issues if CMake is used in top level directory.

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    build:
+      context: ../
+      dockerfile: docker/Dockerfile.gpu
+    ipc: host
+    runtime: nvidia
+    image: nvidia/videoprocessingframework:vpf
+    volumes:
+      - $HOME/.cache:/root/.cache
+    ports:
+      - "127.0.0.1:22911:22911"
```
Comment (philipp-schmidt) [inline docker/docker-compose.yml:15]: What for?

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  vpf:
+    build:
+      context: ../
+      dockerfile: docker/Dockerfile.gpu
+    ipc: host
+    runtime: nvidia
+    image: nvidia/videoprocessingframework:vpf
+    volumes:
+      - $HOME/.cache:/root/.cache
```
Comment (philipp-schmidt) [inline docker/docker-compose.yml:13]: Necessary?

**Review by philipp-schmidt — COMMENTED** [review]
Looks good, nice contrib.

- I wouldn't include docker install scripts though, a link to the docker documentation should be sufficient, stays up-to date and probably includes instructions for windows as well (see my comment below though). [https://docs.docker.com/get-docker/](https://docs.docker.com/get-docker/)
- `docker-compose` is not necessary here. I would advice to keep it cleaner and only provide the command to build and run the docker container itself. Usually a top level `Dockerfile` is the convention here.
- Recent versions of docker use `--gpus` flag instead of nvidia `runtime`, but honestly it's quite confusing. Maybe @sandhawalia can check that.

@rarzumanyan There is currently no reliable way to use GPUs in docker containers under Windows (see [here](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/gpu-acceleration)), so I really don't believe we need Windows instructions here. You Microsoft guys are on your own, good luck ;)

File: docker/docker-compose.yml (hunk 1 above)
Comment (sandhawalia) [reply docker/docker-compose.yml:15]: This is an error. Shall fix.

File: docker/docker-compose.yml (hunk 2 above)
Comment (sandhawalia) [reply docker/docker-compose.yml:13]: Yes this is redundant. Shall fix.

**Comment by sandhawalia** [comment]
I'd recommend to keep `docker-compose` which would allow following flexibility for posterity (If need be)

## Stand alone VPF (+PyTorch)
```
docker-compose -f docker/docker-compose.yaml build vpf
```

## VPF with TensorRT
```
docker-compose -f docker/docker-compose.yaml  build vpf-tensorrt
```

## VPF with OpenCV
```
docker-compose -f docker/docker-compose.yaml  build vpf-opencv
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,46 @@
+SHELL:=/bin/sh
```
Comment (rarzumanyan) [inline Makefile:1]: Any chance you can move Makefile to separate folder? If CMake build is done in root dir, it will overwrite this.

Comment (sandhawalia) [reply Makefile:1]: Yups coming in a few minutes

**Comment by rarzumanyan** [comment]
Creating a feature branch `pr_234` to test PR locally.

**Comment by rarzumanyan** [comment]
@sandhawalia 
Thanks a lot for your contribution!

I've merged the PR to `pr_234` feature branch for local testing. If it goes fine, I'll merge it to `master` later.

**Comment by sandhawalia** [comment]
Great ! Let me know if you run into trouble building / testing. I can work directly from `pr_234` if need be

**Pull request merged**